ha/lock.go

```diff
@@ -0,0 +1,132 @@
+package ha
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad-autoscaler/sdk/helper/uuid"
+)
+
+const (
+	renewalFactor = 0.7
+	waitFactor    = 1.1
+)
+
+type lock interface {
+	Acquire(ctx context.Context, callerID string) (string, error)
+	Release(ctx context.Context) error
+	Renew(ctx context.Context) error
+}
+
+type HAController struct {
+	ID            string
+	renewalPeriod time.Duration
+	waitPeriod    time.Duration
+	randomDelay   time.Duration
+
+	logger log.Logger
+	lock   lock
+}
+
+func NewHAController(l lock, logger log.Logger, lease time.Duration) *HAController {
+	logger = logger.Named("ha_mode")
+
+	rn := rand.New(rand.NewSource(time.Now().Unix())).Intn(100)
+	hac := HAController{
+		lock:          l,
+		logger:        logger,
+		renewalPeriod: time.Duration(float64(lease) * renewalFactor),
+		waitPeriod:    time.Duration(float64(lease) * waitFactor),
+		ID:            uuid.Generate(),
+		randomDelay:   time.Duration(rn) * time.Millisecond,
+	}
+
+	return &hac
+}
+
+func (hc *HAController) Start(ctx context.Context, protectedFunc func(ctx context.Context)) error {
+	hc.logger.Named(hc.ID)
+
+	// To avoid collisions if all the instances start at the same time, wait
+	// a random time before making the first call.
+	hc.wait(ctx)
+
+	waitTimer := time.NewTimer(hc.waitPeriod)
+	defer waitTimer.Stop()
+
+	for {
+		lockID, err := hc.lock.Acquire(ctx, hc.ID)
+		if err != nil {
+			// TODO: What to do with fatal errors?
+			hc.logger.Error("unable to get lock", err)
+		}
+
+		if lockID != "" {
+			hc.logger.Debug("lock acquired, ID", lockID)
+			funcCtx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// Start running the lock protected function
+			go protectedFunc(funcCtx)
+
+			// Maintain lease is a blocking function, will only return in case
+			// the lock is lost or the context is canceled.
+			err := hc.maintainLease(ctx)
+			if err != nil {
+				hc.logger.Debug("lease lost", err)
+				cancel()
+
+				// Give the protected function some time to return before potentially
+				// running it again.
+				hc.wait(ctx)
+			}
+		}
+
+		if !waitTimer.Stop() {
+			<-waitTimer.C
+		}
+		waitTimer.Reset(hc.waitPeriod)
```

Comment:
If the shutdown context fires while we're waiting, I think this blocks shutdown for the wait period? 

In Nomad core we avoid this problem in a few places by starting with a very small wait period (see [`funcs.go#L369-L394`](https://github.com/hashicorp/nomad/blob/main/helper/funcs.go#L369-L394)) and then resetting it to the at the end of the loop as you've done here. Alternately, you could move the `select` up to the top of the loop and initialize the timer to `randomDelay`. That'd let you drop both this and `hc.wait` at the top of this method.

Reply:
For a timer to be successfully reseted, it needs to be stopped and drained first, the stop() function reports if draining is needed but the timer is stoped already:  https://pkg.go.dev/time#Timer.Stop.
I think a ticker is better suited here, Im making the change